OpenGL/Constructs/Geometry.cs

```diff
@@ -52,28 +33,62 @@ public static Vector3[] CalculateNormals(Vector3[] vertexData, int[] elementData
         /// <returns></returns>
         public static Vector3[] CalculateNormals(Vector3[] vertexData, uint[] elementData)
         {
-            Vector3 b1, b2, normal;
+            return CalculateNormals(vertexData.AsSpan(), elementData.AsSpan());
+        }
+
+        /// <summary>
+        /// Calculate the array of vertex normals based on vertex and face information (assuming triangle polygons).
+        /// </summary>
+        /// <param name="vertexData">The vertex data to find the normals for.</param>
+        /// <param name="elementData">The element array describing the order in which vertices are drawn.</param>
+        /// <returns></returns>
+        public static Vector3[] CalculateNormals(ReadOnlySpan<Vector3> vertexData, ReadOnlySpan<uint> elementData)
+        {
             Vector3[] normalData = new Vector3[vertexData.Length];
+            CalculateNormals(vertexData, elementData, normalData);
+            return normalData;
+        }
+
+        /// <summary>
+        /// Calculate the array of vertex normals based on vertex and face information (assuming triangle polygons)
+        /// and put the normals into the provided array.
+        /// </summary>
+        /// <param name="vertexData">The vertex data to find the normals for.</param>
+        /// <param name="elementData">The element array describing the order in which vertices are drawn.</param>
+        /// <param name="normalData">The array the normals will be put into. Has to be the same length as the vertex array</param>
+        /// <returns></returns>
+        public static void CalculateNormals(ReadOnlySpan<Vector3> vertexData, ReadOnlySpan<uint> elementData, Span<Vector3> normalData)
+        {
+            if ((elementData.Length % 3) != 0)
+            {
+                throw new ArgumentOutOfRangeException($"Expected {nameof(elementData)} to be a multiple of 3 as each triangle consists of 3 points.");
+            }
+            if (vertexData.Length != normalData.Length)
+            {
+                throw new ArgumentOutOfRangeException($"Expected {nameof(vertexData)} and {nameof(normalData)} to have the same length.");
+            }
 
-            for (int i = 0; i < elementData.Length / 3; i++)
+            for (int i = 0; i < elementData.Length; i += 3)
             {
-                uint cornerA = elementData[i * 3];
-                uint cornerB = elementData[i * 3 + 1];
-                uint cornerC = elementData[i * 3 + 2];
+                int cornerAIndex = (int)elementData[i + 0];
```

Comment:
Any reason for changing to use `int` with a cast instead of just indexing the vertexData with `uint`?